### PR TITLE
Don't recreate whole completion item upon calculation of properties.

### PR DESCRIPTION
**Pull request by ToddGrun**
Found this while looking at customer completion trace. It's a very small improvement (0.2% of CPU during completion time period in customer profile), but it's simple enough to wait to create the completion item until we've determined the properties.

image

The general usage of the ImmutableDictionary for properties is where the bigger cost liess, but I'm not seeing a way around that is it's part of the Public API.

### CODE REVIEW

**Comment by ToddGrun** [comment]
Changing this to draft as I found another place in completion using SegmentedLists and am thinking this might require more thought.

File: src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs [hunk 1]
```diff
@@ -66,22 +68,21 @@ internal static class SymbolCompletionItem
...
-                .AddProperty("SymbolName", symbol.Name);
+            properties = properties
+                .Add("SymbolKind", ((int)symbol.Kind).ToString())
+                .Add("SymbolName", symbol.Name);
 
-            return isGeneric ? item.AddProperty("IsGeneric", isGeneric.ToString()) : item;
+            return isGeneric ? properties.Add("IsGeneric", isGeneric.ToString()) : properties;
         }
 
         public static CompletionItem AddShouldProvideParenthesisCompletion(CompletionItem item)
```
[the remark this answers is not in the document]
Comment (ToddGrun) [reply src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs:88]: Unfortunately, that looks a fair bit more complicated as the code that does that lies in SymbolCompletionProvider

File: src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs [hunk 2]
```diff
@@ -50,6 +50,8 @@ internal static class SymbolCompletionItem
             }
 
             props = props.Add("ContextPosition", contextPosition.ToString());
+            props = WithSupportedPlatforms(props, supportedPlatforms);
```
Comment (genlu) [inline src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs:53]: Given that we are now constructing the props first, we can use ImmutableDictionary.Builder to save a couple more allocations for each item creation

File: src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs [hunk 3]
```diff
@@ -66,22 +68,21 @@ internal static class SymbolCompletionItem
...
                 isComplexTextEdit: isComplexTextEdit);
 
-            item = WithSupportedPlatforms(item, supportedPlatforms);
-            return symbolEncoder(symbols, item);
+            return item;
         }
 
-        public static CompletionItem AddSymbolEncoding(IReadOnlyList<ISymbol> symbols, CompletionItem item)
-            => item.AddProperty("Symbols", EncodeSymbols(symbols));
+        public static ImmutableDictionary<string, string> AddSymbolEncoding(IReadOnlyList<ISymbol> symbols, ImmutableDictionary<string, string> properties)
```
Comment (genlu) [inline src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs:74]: Can this be private? (same as `AddSymbolInfo` nelow)

**Review by genlu — APPROVED** [review]
LGTM beside comments. I can take a look at fixing `AddShouldProvideParenthesisCompletion` oncet his is merged.

File: src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs (hunk 3 above)
Comment (ToddGrun) [reply src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs]: Done

File: src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs (hunk 2 above)
Comment (ToddGrun) [reply src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs]: Good idea!

**Pull request merged**